pkg/threescale/connectors/backend.go

```diff
@@ -0,0 +1,141 @@
+package connectors
+
+import (
+	"fmt"
+	"time"
+
+	backend "github.com/3scale/3scale-go-client/client"
+	"github.com/3scale/3scale-istio-adapter/pkg/threescale/metrics"
+)
+
+type AuthRepResult string
+
+const Unknown AuthRepResult = "UNKNOWN"
+const Ok AuthRepResult = "AUTHENTICATED"
+const Unauthenticated AuthRepResult = "UNAUTHENTICATED"
+const cacheMiss AuthRepResult = "UNCACHED"
+
+const (
+	UserKey = iota + 1
+	AppId
+)
+
+type authRepFn func(auth backend.TokenAuth, key string, svcID string, params backend.AuthRepParams, extension map[string]string) (backend.ApiResponse, error)
+
+type AuthRepUserAuth struct {
+	// Api Key Authn Pattern
+	UserKey string
+	// App Id Authn Pattern - Mutually exclusive with Api Key Pattern
+	AppID string
+	// Optional App Key as a key pair for App ID pattern
+	AppKey   string
+	AuthType int
+}
+
+type AuthRepRequest struct {
+	ServiceID string
+	UserAuth  AuthRepUserAuth
+	TokenAuth backend.TokenAuth
+	Metrics   backend.Metrics
+}
+
+type authRepRemoteCall struct {
+	authKey        string
+	params         backend.AuthRepParams
+	callback       authRepFn
+	remoteHostname string
+}
+
+type BackendCache interface {
+	AuthRep(svcID string, metrics backend.Metrics) AuthRepResult
+	Set()
+}
+
+type BackendConnector struct {
+	cache           BackendCache
+	reportMetricsFn metrics.ReportMetricsFn
+}
+
+type LocalCache struct {
+}
+
+func (l LocalCache) AuthRep(svcID string, metrics backend.Metrics) AuthRepResult {
+	return cacheMiss
+}
+
+func (l LocalCache) Set() {
+
+}
+
+func NewBackendConnector() *BackendConnector {
+	return &BackendConnector{}
+}
+
+func (bc *BackendConnector) DoAuthRep(req AuthRepRequest, metrics backend.Metrics, c *backend.ThreeScaleClient) (backend.ApiResponse, error) {
+	var resp backend.ApiResponse
+
+	callRemote := func() (backend.ApiResponse, error) {
+		remoteReq, err := bc.authRepRemoteReqBuilder(req.UserAuth, metrics, c)
+		if err != nil {
+			return resp, err
+		}
+
+		return bc.callRemote(req.TokenAuth, remoteReq, req.ServiceID)
+	}
```

Comment:
Is this callback mechanism really needed here? I know it was already there (moved code), but couldn't this just be done in the `cacheMiss` case specifying this data to the function invoking a 3scale SM API (backend) call?

Reply:
Well cache miss and also caching not enabled, in both cases we need to determine what api in backend we need to call so producing duplicate code in both cases, which is why I took this approach.

Thinking about how to restructure to reduce complexity.